File: src/Aether.Storage/InMemoryKeyValueStore.cs
// SPDX-License-Identifier: MIT

using System.Collections.Concurrent;
using System.Runtime.CompilerServices;

namespace Aether.Storage;

/// <summary>
/// Process-local, volatile key-value store backed by <see cref="ConcurrentDictionary{TKey, TValue}"/>.
/// Suitable for tests and demos. Loses everything on process exit.
/// </summary>
public sealed class InMemoryKeyValueStore : IKeyValueStore
{
    private readonly ConcurrentDictionary<string, byte[]> _entries = new(StringComparer.Ordinal);

    public Task<byte[]?> GetAsync(string key, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        _entries.TryGetValue(key, out var value);
        return Task.FromResult(value);
    }

    public Task PutAsync(string key, byte[] value, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        ArgumentNullException.ThrowIfNull(value);
        // Defensive copy so the caller can't mutate the stored bytes.
        var copy = new byte[value.Length];
        Buffer.BlockCopy(value, 0, copy, 0, value.Length);
        _entries[key] = copy;
        return Task.CompletedTask;
    }

    public Task<bool> RemoveAsync(string key, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        return Task.FromResult(_entries.TryRemove(key, out _));
    }

    public Task<bool> ContainsAsync(string key, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(key);
        return Task.FromResult(_entries.ContainsKey(key));
    }

    public async IAsyncEnumerable<string> ListKeysAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        foreach (var key in _entries.Keys)
        {
            cancellationToken.ThrowIfCancellationRequested();
            yield return key;
            await Task.Yield();
        }
    }
}

File: typescript/src/storage/InMemoryKeyValueStore.ts
/**
 * Process-local, volatile {@link KeyValueStore} backed by a {@link Map}.
 * Suitable for tests and demos. Loses everything on process exit.
 *
 * Mirrors the C# {@code InMemoryKeyValueStore} in
 * src/Aether.Storage/InMemoryKeyValueStore.cs.
 *
 * SPDX-License-Identifier: MIT
 */
import { KeyValueStore } from "./KeyValueStore.js";

export class InMemoryKeyValueStore implements KeyValueStore {
  private readonly entries: Map<string, Uint8Array> = new Map();

  async get(key: string): Promise<Uint8Array | null> {
    if (!key) throw new Error("key cannot be empty");
    const value = this.entries.get(key);
    if (value === undefined) return null;
    // Defensive copy on the way out — caller can't mutate stored bytes.
    return new Uint8Array(value);
  }

  async put(key: string, value: Uint8Array): Promise<void> {
    if (!key) throw new Error("key cannot be empty");
    if (value === null || value === undefined) {
      throw new Error("value cannot be null/undefined");
    }
    // Defensive copy on the way in — caller can't subsequently mutate
    // the stored bytes.
    this.entries.set(key, new Uint8Array(value));
  }

  async remove(key: string): Promise<boolean> {
    if (!key) throw new Error("key cannot be empty");
    return this.entries.delete(key);
  }

  async contains(key: string): Promise<boolean> {
    if (!key) throw new Error("key cannot be empty");
    return this.entries.has(key);
  }

  async listKeys(prefix?: string): Promise<string[]> {
    const all = Array.from(this.entries.keys());
    if (!prefix) return all;
    return all.filter((k) => k.startsWith(prefix));
  }
}
